Ubs/ refactoring address management creating a dedicated service and controller 8124

Andrii-Danylenko opened the pull request:
## GitHub issue

* [GitHub ticket](https://github.com/ita-social-projects/GreenCity/issues/8124)

## Summary of issue

To improve code maintainability, scalability, and separation of concerns, the address management logic should be refactored into a dedicated service. Currently, address-related operations are handled within existing services, making the codebase more complex and harder to maintain.

Scope of Work:

Extract all address-related logic into a separate AddressService.
Develop a new AddressController to handle address-specific operations via dedicated endpoints.
Ensure that other services interact with addresses through the new service instead of handling address logic directly.
Refactor existing code to integrate the new service properly.
Implement necessary unit and integration tests to verify the functionality of the new structure.
This refactoring will enhance system modularity, simplify future enhancements, and improve overall maintainability.

## Summary of change

All functionality related to addresses was moved to a separate class AddressService and AddressController. Old tests were also moved and new ones were written. Coverage level - 96%

## Testing approach

Unit testing and manual testing


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a new endpoint to retrieve address details by order identifier.
  - Expanded address management capabilities, allowing users to update, save, and delete addresses.
  - Introduced functionality to fetch district information based on region or city.
  - Address data now includes a status indicator to clarify its current state.
  - Added utility methods for creating predefined address request DTOs for testing.
  - Enhanced test coverage for address-related functionalities, including equality checks for address DTOs.

- **Bug Fixes**
  - Improved error handling for address retrieval and updates, ensuring appropriate exceptions are thrown for non-existent orders.

- **Refactor**
  - Consolidated address management logic into a unified service, streamlining operations and enhancing validation and error handling.
  - Removed outdated methods related to address management from various services, focusing on the new address service.
  - Updated test cases to reflect changes in the address management logic and service interactions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Andrii-Danylenko committed:
Moved functionality related to addresses to a separate AddressService

Andrii-Danylenko committed:
Added tests for AddressServiceImpl

Andrii-Danylenko committed:
Formatted files

Andrii-Danylenko committed:
Fixed AddressControllerTest

Cr1stal423 commented on core/src/main/java/greencity/controller/AddressController.java:
```diff
@@ -244,4 +243,26 @@ public ResponseEntity<Void> updateAddress(@RequestBody @Valid UpdateAddressDto a
+        @ApiResponse(responseCode = "403", description = HttpStatuses.FORBIDDEN, content = @Content),
+        @ApiResponse(responseCode = "404", description = HttpStatuses.NOT_FOUND, content = @Content)
+    })
+    @GetMapping("/read-address-order/{id}")
```
In ManagementOrderController, the endpoint used to be from @RequestMapping("/ubs/management"), you need to either make these changes here as well or agree with the frontend to change the endpoint path

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +60,309 @@ public class AddressServiceImpl implements AddressService {
+     * {@inheritDoc}
+     */
+    @Override
+    @jakarta.transaction.Transactional
```
Import this annotation instead of using manual packaging

Cr1stal423 replied:
Perhaps you should use the transaction from the org.springframework.transaction.annotation.Transactional package? They are already imported

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +60,309 @@ public class AddressServiceImpl implements AddressService {
+     * {@inheritDoc}
+     */
+    @Override
+    @jakarta.transaction.Transactional
```
also here

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +60,309 @@ public class AddressServiceImpl implements AddressService {
+     * {@inheritDoc}
+     */
+    @Override
+    @jakarta.transaction.Transactional
```
and here

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +60,309 @@ public class AddressServiceImpl implements AddressService {
+     * {@inheritDoc}
+     */
+    @Override
+    @jakarta.transaction.Transactional
```
and here

Cr1stal423 commented on service/src/test/java/greencity/service/ubs/AddressServiceTest.java:
```diff
@@ -1,29 +1,79 @@
+import greencity.repository.*;
+
+import java.lang.reflect.Method;
+import java.util.*;
```
remove wildcard imports

Cr1stal423 commented on service-api/src/main/java/greencity/service/ubs/AddressService.java:
```diff
@@ -1,6 +1,15 @@
+import greencity.dto.address.AddressDto;
 import greencity.dto.address.UpdateAddressDto;
+import greencity.dto.location.api.DistrictDto;
+import greencity.dto.order.*;
```
wildcard import

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -1,20 +1,56 @@
 import greencity.dto.address.UpdateAddressDto;
+import greencity.dto.location.api.DistrictDto;
+import greencity.dto.location.api.LocationDto;
+import greencity.dto.order.*;
```
wildcard import

Cr1stal423 commented:
Also fix SonarQube issue
![image](https://github.com/user-attachments/assets/edd969ea-7501-47ae-88c1-d3b58e770816)

Andrii-Danylenko committed:
Minor code refactoring

Andrii-Danylenko committed:
Minor code refactoring

Andrii-Danylenko committed:
Minor code refactoring

Andrii-Danylenko committed:
Minor code refactoring

Cr1stal423 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +70,309 @@ public class AddressServiceImpl implements AddressService {
+        return deletedAddress.orElse(null);
+    }
+
+    private boolean areAddressesEqual(CreateAddressRequestDto a, CreateAddressRequestDto b) {
```
Does it make sense to write this method instead of using implementation of equals?

Andrii-Danylenko replied:
Legacy code. Haven't change it yet

Andrii-Danylenko replied:
After careful research, I concluded that the logic uses two versions of equals. Globally for the application, it uses the equals of the class, and in the AddressService implementation it uses its own, which is based on the logic for checking duplicates for individual fields. It must be left as is

Cr1stal423 commented on core/src/main/java/greencity/controller/AdminUbsController.java:
```diff
@@ -13,10 +13,7 @@
-import greencity.service.ubs.OrdersForUserService;
-import greencity.service.ubs.ValuesForUserTableService;
-import greencity.service.ubs.ViolationService;
+import greencity.service.ubs.*;
```
U miss this wildcard import

dxrknesss replied:
@Andrii-Danylenko, I suggest you just 
`mvn formatter:format clean package` , so that it will format your code and show if there are any checkstyle errors.

Cr1stal423 replied:
And also in the IDE settings you can set 99 impots to use imports from *
![image](https://github.com/user-attachments/assets/a70c3e15-76c2-495c-9241-f528e48e3bb8)

VitaliiChernenko replied:
@Cr1stal423, Oh, good idea. I'll also add these settings to my IDE. Thanks.

dxrknesss commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +70,309 @@ public class AddressServiceImpl implements AddressService {
+            return false;
+        }
+
+        return (Objects.equals(a.getRegion(), b.getRegion()) || Objects.equals(a.getRegionEn(), b.getRegionEn()))
+            && (Objects.equals(a.getCity(), b.getCity()) || Objects.equals(a.getCityEn(), b.getCityEn()))
+            && (Objects.equals(a.getDistrict(), b.getDistrict())
+                || Objects.equals(a.getDistrictEn(), b.getDistrictEn()))
+            && Objects.equals(a.getHouseNumber(), b.getHouseNumber())
+            && Objects.equals(a.getEntranceNumber(), b.getEntranceNumber())
+            && Objects.equals(a.getHouseCorpus(), b.getHouseCorpus());
```
maybe it's better to encapsulate this method in CreateAddressRequestDto object, you won't be calling so many getters, and when other part of code will need to compare addresses like this, it won't depend on this service, but rather on the object itself.

ChernenkoVitaliy committed:
Added endpoint for receiving total number of orders for employees

Added method getOrdersCountByTariffs for receiving total numbers of orders by list of tariffs in the BigOrderTableRepository; (#1616)
Added tests for getOrdersCountByTariffs method in the BigOrderTableRepositoryTest;
Added method getTotalNumberOfOrdersByEmployee in the BigOrderTableServiceView and created realisation for it in the BigOrderTableViewServiceImpl;
Added unit tests for method getTotalNumberOfOrdersByEmployee into BigOrderTableViewServiceImplTest;
Renamed class BigOrderTableServiceImplTest as BigOrderTableViewServiceImplTest;
Removed old realisation of receiving total number of orders from UBSManagementService, UBSManagementServiceImpl and UBSManagementServiceImplTest;
Refactored method getOrdersCount in the ManagementOrderController according to new requerements;
Refactored tests in the ManagementOrderControllerTest according to changes in the controller;

ChernenkoVitaliy committed:
Removed import wildcards, refactored tests

Andrii-Danylenko committed:
Removed import wildcards, refactored tests

Andrii-Danylenko committed:
Merge branch 'dev' into UBS/-Refactoring-Address-Management-Creating-a-Dedicated-Service-and-Controller-8124

Andrii-Danylenko committed:
Removed import wildcards, refactored tests

Andrii-Danylenko committed:
Removed import wildcards, refactored tests

Andrii-Danylenko committed:
Removed import wildcards, refactored tests

Andrii-Danylenko committed:
Removed import wildcards, refactored tests

maks741 commented on service-api/src/main/java/greencity/dto/CreateAddressRequestDto.java:
```diff
@@ -75,4 +76,18 @@ public class CreateAddressRequestDto {
     @NotEmpty(message = notEmptyValidationMessage)
     private String streetEn;
+
+    public boolean areAddressesEqual(CreateAddressRequestDto otherAddress) {
```
shouldn't that be an override of equals method? Currently, there are scenarios when 'equals' generated by Lombok and 'areAddressesEqual' produce different output for same input

Andrii-Danylenko replied:
The application uses two versions of equals. In some places equality is checked completely, and in some places only equality at address is checked. Changing equals will break the program.

maks741 commented on service-api/src/test/java/greencity/dto/address/CreateAddressRequestDtoTest.java:
```diff
@@ -0,0 +1,35 @@
+
+public class CreateAddressRequestDtoTest {
+    @Test
+    void testAreAddressesEqualByCityAndRegionAndDistrict() {
```
works fine, but would be great if variable names were more descriptive and each assertion would be extracted into a unit test describing that exactly is being validated. I know it may be difficult to come up with values to create CreateAddressRequestDto for it to be more descriptive, but sure ChatGPT can help you with that

maks741 commented on service/src/main/java/greencity/service/ubs/AddressServiceImpl.java:
```diff
@@ -24,8 +70,295 @@ public class AddressServiceImpl implements AddressService {
+                .areAddressesEqual((modelMapper.map(address, CreateAddressRequestDto.class))))
+            .findFirst();
+
+        return deletedAddress.orElse(null);
```
why not return the Optional? The 'ifPresentOrElse' method of Optional class fits really well instead of null checks that you currently perform for each call of the 'checkIfAddressExist' method

Andrii-Danylenko replied:
Great idea. Will change return type to Optional.empty

Andrii-Danylenko committed:
Replaced null checks with Optional class

Cr1stal423 commented:
I can't leave a comment on the code itself, but here you have put certain variables as constants, for them you should use SCREAMING_SNAKE_CASE. Also, remove the access modifier from the test class
![image](https://github.com/user-attachments/assets/520fafb0-8cf5-43da-a98f-9b3d9311a670)

Andrii-Danylenko committed:
Minor code refactoring

Cr1stal423 commented:
Resolve conflicts with dev branch

Andrii-Danylenko committed:
Merge branch 'dev' into UBS/-Refactoring-Address-Management-Creating-a-Dedicated-Service-and-Controller-8124

Andrii-Danylenko committed:
Fixed tests (they were merged incorrectly)
